pkg/identity/nonglobal/temp_id.go

```diff
@@ -0,0 +1,178 @@
+// identities. Temporary security identities are used by new pods before a
+// global security identity is allocated by cilium-operator.
+type TempSecIDAllocator struct {
+	allocator          *basicallocator.BasicIDAllocator
+	tempIDCache        *TempIDCache
+	endpointListerFunc func() []*endpoint.Endpoint
+}
+
+// NewTempSecIDAllocator creates a new TempSecIDAllocator for ID values
+// between the provided minIDValue and maxIDValue.
+func NewTempSecIDAllocator(endpointListerFunc func() []*endpoint.Endpoint) *TempSecIDAllocator {
+	return &TempSecIDAllocator{
+		allocator:          basicallocator.NewBasicIDAllocator(identity.DefaultMinTempID, identity.DefaultMaxTempID),
+		tempIDCache:        NewTempIDCache(),
+		endpointListerFunc: endpointListerFunc,
+	}
+}
+
+type TempIDCache struct {
+	mu                lock.RWMutex
+	idToIdentity      map[identity.NumericIdentity]*identity.Identity
+	keyToIdentity     map[string]*identity.Identity
+	markedForDeletion map[identity.NumericIdentity]bool
+}
+
+func NewTempIDCache() *TempIDCache {
+	return &TempIDCache{
+		mu:                lock.RWMutex{},
+		idToIdentity:      make(map[identity.NumericIdentity]*identity.Identity),
+		keyToIdentity:     make(map[string]*identity.Identity),
+		markedForDeletion: make(map[identity.NumericIdentity]bool),
+	}
+}
+
+// FindOrCreateTempID gets ans existing or creates a temporary security identity
+// for the specified labels, to be used before operator creates a global
+// security identity.
+func (a *TempSecIDAllocator) FindOrCreateTempID(lbls labels.Labels) (*identity.Identity, error) {
+	if id, exists := a.LookupByIDKey(&key.GlobalIdentity{LabelArray: lbls.LabelArray()}); exists {
+		return id, nil
+	}
+
+	allocatedID, err := a.allocator.AllocateRandom()
+	if err != nil {
+		return nil, err
+	}
+
+	numID := identity.NumericIdentity(int64(allocatedID))
+	id := identity.NewIdentity(numID, lbls)
+	a.insertToCache(id)
+
+	return id, nil
+}
+
+func (a *TempSecIDAllocator) LookupByID(numID identity.NumericIdentity) (*identity.Identity, bool) {
+	if !a.allocator.IsInPoolRange(idpool.ID(numID)) {
+		return nil, false
+	}
+
+	cache := a.tempIDCache
+	cache.mu.RLock()
+	defer cache.mu.RUnlock()
+
+	id, exists := cache.idToIdentity[numID]
+	return id, exists
+}
+
+func (a *TempSecIDAllocator) LookupByIDKey(idKey *key.GlobalIdentity) (*identity.Identity, bool) {
+	cache := a.tempIDCache
+	cache.mu.RLock()
+	defer cache.mu.RUnlock()
+
+	id, exists := cache.keyToIdentity[idKey.GetKey()]
+	return id, exists
+}
+
+func (a *TempSecIDAllocator) insertToCache(id *identity.Identity) {
+	numID := id.ID
+	if !a.allocator.IsInPoolRange(idpool.ID(numID)) {
+		return
+	}
+
+	cache := a.tempIDCache
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	idKey := &key.GlobalIdentity{LabelArray: id.LabelArray}
+	cache.idToIdentity[numID] = id
+	cache.keyToIdentity[idKey.GetKey()] = id
+}
+
+func (a *TempSecIDAllocator) StartPeriodicGC(stopChan chan struct{}) {
```

Comment:
Can you add a comment what GC suppose to do?

Reply:
I added a comment below to the runGC.